File: tests/test_data/autotasks.py
from datetime import datetime

from ichrisbirch.models import AutoTask

BASE_DATA: list[AutoTask] = [
    AutoTask(
        name='AutoTask 1 Chore with notes priority 5 not completed',
        notes='Notes for task 1',
        category='Chore',
        priority=5,
        frequency='Daily',
        first_run_date=datetime(2020, 4, 20, 3, 3, 39, 50648).isoformat(),
        last_run_date=datetime(2020, 4, 24, 3, 3, 39, 50648).isoformat(),
        run_count=5,
    ),
    AutoTask(
        name='AutoTask 2 Home without notes priority 10 not completed',
        notes=None,
        category='Home',
        priority=10,
        frequency='Weekly',
        first_run_date=datetime(2020, 3, 20, 3, 3, 39, 50648).isoformat(),
        last_run_date=datetime(2020, 3, 24, 3, 3, 39, 50648).isoformat(),
        run_count=1,
    ),
    AutoTask(
        name='AutoTask 3 Home with notes priority 15 completed',
        notes='Notes for task 3',
        category='Home',
        priority=15,
        frequency='Quarterly',
        first_run_date=datetime(2020, 1, 20, 3, 3, 39, 50648).isoformat(),
        last_run_date=datetime(2020, 1, 24, 3, 3, 39, 50648).isoformat(),
        run_count=2,
    ),
]

File: cli/internal/api/autotasks.go
package api

import (
	"context"
	"fmt"
	"net/url"
	"time"
)

// AutoTask mirrors the autotasks JSON — a recurring task template the scheduler
// uses to spawn tasks. It is read-only from the CLI: the scheduler owns the
// run bookkeeping (first_run_date/last_run_date/run_count). notes is nullable.
type AutoTask struct {
	ID            int       `json:"id"`
	Name          string    `json:"name"`
	Category      string    `json:"category"`
	Priority      int       `json:"priority"`
	Notes         *string   `json:"notes"`
	Frequency     string    `json:"frequency"`
	MaxConcurrent int       `json:"max_concurrent"`
	FirstRunDate  time.Time `json:"first_run_date"`
	LastRunDate   time.Time `json:"last_run_date"`
	RunCount      int       `json:"run_count"`
}

// ListAutoTasks returns recurring task templates, most recently run first
// (GET /autotasks/). A nil limit fetches all; a non-nil limit caps the count.
func (c *Client) ListAutoTasks(ctx context.Context, limit *int) ([]AutoTask, error) {
	params := url.Values{}
	applyLimit(params, limit)
	var autotasks []AutoTask
	if err := c.get(ctx, withQuery("/autotasks/", params), &autotasks); err != nil {
		return nil, err
	}
	return autotasks, nil
}

// GetAutoTask returns a single recurring task template (GET /autotasks/{id}/).
// Missing is 404.
func (c *Client) GetAutoTask(ctx context.Context, id int) (AutoTask, error) {
	var autotask AutoTask
	if err := c.get(ctx, fmt.Sprintf("/autotasks/%d/", id), &autotask); err != nil {
		return AutoTask{}, err
	}
	return autotask, nil
}
